src/pages/index.js

```diff
@@ -13,53 +15,181 @@ import {
+  });
+
+function createCard(cardData) {
+  const card = new Card(cardData, "#card-template", {
+    handleImageClick: () => imagePopup.open(cardData.name, cardData.link),
+    handleCardDelete,
+    handleCardLike: handleCardLike,
+  });
+  return card.getView();
+}
+
+function renderCard(cardData) {
+  const cardElement = createCard(cardData);
+  section.addItem(cardElement);
+}
+
+// form validations
+const changeProfileAvatarForm = document.forms["avatar-form"];
 const editProfileForm = document.forms["profile-form"];
+const addCardForm = document.forms["card-form"];
+
+const changeProfileAvatarFormValidator = new FormValidator(
+  options,
+  changeProfileAvatarForm
+);
 const editProfileFormValidator = new FormValidator(options, editProfileForm);
+const addCardFormValidator = new FormValidator(options, addCardForm);
+
+changeProfileAvatarFormValidator.enableValidation();
+addCardFormValidator.enableValidation();
 editProfileFormValidator.enableValidation();
 
+// modal instances
 const imagePopup = new PopupWithImage("#preview-image-modal");
-imagePopup.setEventListeners();
+const changeProfileAvatarPopup = new PopupWithForm(
+  "#change-profile-avatar-modal",
+  handleChangeProfileAvatarFormSubmit
+);
+const editProfilePopup = new PopupWithForm(
+  "#edit-modal",
+  handleEditProfileFormSubmit
+);
+const addCardPopup = new PopupWithForm(
+  "#add-card-modal",
+  handleAddCardFormSubmit
+);
+const deleteCardPopup = new PopupWithConfirmation("#delete-card-modal");
 
-const addCardPopup = new PopupWithForm("#add-card-modal", (value) => {
-  const cardData = {
-    name: value.Title,
-    link: value.URL,
-  };
-  renderCard(cardData);
-  addCardPopup.close();
-});
-addCardPopup.setEventListeners();
+// handle functions
+function handleCardDelete(cardId) {
+  deleteCardPopup.open();
+  deleteCardPopup.setSubmitAction(() => {
+    deleteCardPopup.setLoading(true, "Deleting...");
+    api
+      .deleteCard(cardId._id)
+      .then(() => {
+        deleteCardPopup.close();
+        cardId.removeCard();
+      })
+      .catch((err) => {
+        console.error(err);
+      })
+      .finally(() => {
+        deleteCardPopup.setLoading(false, "Yes");
+      });
+  });
+}
 
-const userInfo = new UserInfo(".profile__title", ".profile__description");
+function handleChangeProfileAvatarFormSubmit(url) {
+  changeProfileAvatarPopup.setLoading(true, "saving...");
+  api
+    .updateAvatar(url)
+    .then((data) => {
+      userInfo.setAvatar(data.avatar);
+      changeProfileAvatarPopup.close();
+    })
+    .catch((err) => {
+      console.error(err);
+    })
+    .finally(() => changeProfileAvatarPopup.setLoading(false, "Save"));
+}
 
-const editProfilePopup = new PopupWithForm("#edit-modal", (profileData) => {
-  userInfo.setUserInfo(profileData.name, profileData.description);
-  editProfilePopup.close();
-});
-editProfilePopup.setEventListeners();
+function handleEditProfileFormSubmit(data) {
+  editProfilePopup.setLoading(true, "Saving...");
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```